<launch>
    <!--  ##############  Input arguments  ##############  -->
    <arg name="enableSafety" default="true"/>
    <arg name="enableJoystick" default="false"/>
    <arg name="enableKeyboard" default="false"/>
    <arg name="joystickDevice" default="/dev/input/js0"/>
    <arg name="enableProfiler" default="false"/>
    <arg name="autosave_params_path" default="$(find cola2_hug)/config/saved_configurations"/>

    <!--  ##############  LOAD NODES  ##############  -->
    <!-- Robot description hug -->
    <node name="robot_state_publisher_hug" pkg="robot_state_publisher" type="state_publisher" >
        <remap from="robot_description" to="robot_description_hug"/>
    </node>

    <!-- Safety nodes -->
    <node name="param_logger" pkg="cola2_safety" type="param_logger.py" output="screen" args="$(arg autosave_params_path)"/>
    <node name="up_time" pkg="cola2_safety" type="up_time.py" output="screen"/>
    <node name="safe_depth_altitude" pkg="cola2_safety" type="safe_depth_altitude.py" output="screen"/>
    <node name="set_zero_velocity" pkg="cola2_safety" type="set_zero_velocity.py" output="screen"/>
    <node name="recovery_actions" pkg="cola2_safety" type="recovery_actions.py" output="screen"/>
    <node name="virtual_cage" pkg="cola2_safety" type="virtual_cage.py" output="screen"/>
    <node pkg="diagnostic_aggregator" type="aggregator_node" name="diagnostic_aggregator" />
    <node if="$(arg enableSafety)" name="safety_supervisor" pkg="cola2_safety" type="safety_supervisor.py"/>
    <node name="computer_logger" pkg="cola2_safety" type="computer_logger.py" output="screen"/>
    <node name="vehicle_status_parser" pkg="cola2_hug" type="vehicle_status_parser.py" respawn="true" output="screen"/>

    <!-- Run dynamics -->
    <node name="dynamics" pkg="cola2_sim" type="dynamics.py"/>
    <node name="sim_nav_sensors_hug" pkg="cola2_hug" type="sim_nav_sensors_hug.py" output="screen"/>
    <node name="sim_actuators_hug" pkg="cola2_hug" type="sim_actuators_hug.py" output="screen"/>

    <!-- Run Navigation Stack -->
    <node name="navigator" pkg="cola2_hug" type="navigator_hug" respawn="false" output="screen"/>

    <!-- Run Control Stack -->
    <node name="teleoperation" pkg="cola2_control" type="teleoperation.py" output="screen"/>
    <node name="controller" pkg="cola2_hug" type="controller_hug" respawn="true"/>
    <node name="captain" pkg="cola2_control" type="captain" respawn="false" output="screen"/>
	  <node name="pilot" pkg="cola2_control" type="pilot" respawn="false" output="screen"/>

    <!--  #################  Other nodes  #################  -->
    <node name="bag_node" pkg="cola2_lib" type="bag_node.py" output="screen"/>
    <node name="mission_wait" pkg="cola2_lib" type="mission_wait.py" output="screen"/>
    <node name="test_thrusters" pkg="cola2_safety" type="test_thrusters.py" output="screen"/>

    <group if="$(arg enableProfiler)">
      <node name="sim_profiler" pkg="cola2_sim" type="sim_profiler.py" output="screen">
          <param name="frame_id" value="hug"/>
          <!-- If these parameters change, the hug.urdf must change too -->
          <param name="tf_x" value="0.7"/>
          <param name="tf_y" value="0.0"/>
          <param name="tf_z" value="-0.15"/>
          <param name="tf_roll" value="0.0"/>
          <param name="tf_pitch" value="0.0"/>
          <param name="tf_yaw" value="3.14159"/>
      </node>
    </group>

    <!-- Run Teleoperation -->
    <group if="$(arg enableJoystick)">
        <node respawn="true" pkg="joy" type="joy_node" name="joystick">
            <param name="dev" value="$(arg joystickDevice)"/>
            <param name="deadzone" value="0.05"/>
        </node>
        <node respawn="false" pkg="cola2_hug" type="logitech_fx10_to_teleoperation_hug.py" name="fx10"/>
    </group>
    <group if="$(arg enableKeyboard)">
        <node name="keyboard" pkg="cola2_control" type="keyboard" output="screen">
            <remap from="/keyboard" to="/joy"/>
        </node>
        <node name="keyboard_to_teleoperation" pkg="cola2_control" type="keyboard_to_teleoperation.py" />
    </group>
</launch>
